### PR TITLE
Add types to react-emotion

### PR DESCRIPTION
**Pull request by renatorib**
**What**:
TypeScript declarations for react-emotion package

**Why**:
Support people who use TypeScript and need typings. 

**How**:
* Created `react-emotion.d.ts` and linked in "types" package.json field
* Followed same structure/names from #379 

**Checklist**:
- [ ] Documentation
- [x] Tests
- [x] Code complete

---

### Supports

**Props Inference by Tag Name:**

image

---

**Supports template string and object as styles:**

image

---

**Passing Custom Props:**

image
*Note1: needs a second type parameter passing tag name because it stop to infer when you pass props at first parameter*
*Note2: glamorous do not support custom props with shorthands, like Link2 :stuck_out_tongue:. It was a little tricky, but works...* 

---

**Create styled with another component:**

image

---

**Pass custom props to style another component:**

image
*Note3: as well as Note1 it's not infer 'MyComponent's Props' when you pass props as first type parameter, so you need to pass 'MyComponent' props together*
```tsx
type Props = MyNewProps & MyComponentProps
```
---

**Theming props:**

By default props.theme is `any`, but you can create another styled instance passing theme props.
Basically the same approach from styled-components 

image

---

**withComponent:**

image

Also works passing another component instead of tag name

### CODE REVIEW

File: packages/react-emotion/typescript_tests/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
Comment (luke-john) [inline packages/react-emotion/typescript_tests/tsconfig.json]: For glamorous we've added declaration: true to help make sure we're exporting everything required for use in libraries Microsoft/TypeScript#5938.

**Comment by luke-john** [comment]
Awesome work doing these types. They're looking great 👍 .

File: packages/react-emotion/typings/react-emotion.d.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import { StatelessComponent, Component as ClassComponent, CSSProperties } from 'react'
+import { Interpolation as EmotionInterpolation } from 'emotion'
+
+export type InterpolationFn<Props = {}> = (
+  props: Props
+) => EmotionInterpolation
+
+export type Interpolation<Props = {}> =
+  | InterpolationFn<Props>
+  | EmotionInterpolation
```
Comment (mitchellhamilton) [inline packages/react-emotion/typings/react-emotion.d.ts]: I don't know much about typescript but I don't think this handles the fact that there can be nested function interpolations and they will receive props.

```jsx
const SomeComponent = styled.div`
  display: ${(p) => props => props.display};
  color: ${[props => 'hotpink']};
`
```

**Comment by screendriver** [comment]
But you still need Babel in the build pipeline as I already asked here #163?

File: packages/react-emotion/typings/react-emotion.d.ts (hunk 2 above)
Comment (renatorib) [reply packages/react-emotion/typings/react-emotion.d.ts]: It works on emotion?
props => props_ => ...

It doesn't make any sense to me. What's this use case for?

**Comment by renatorib** [comment]
@screendriver This is only typings, do not deal with build, or any emotion code. I don't even know the internals of emotion yet.

**Comment by screendriver** [comment]
Ah ok. But TypeScript definitions for emotion are a little but useless when you can't use emotion without Babel 😉 (emotion only works as Babel plugin so Babel is mandatory) 

Apart from that: thank you for the typings 😎👍

**Comment by renatorib** [comment]
If you don't want to use babel (or can't) just don't use it. Pick another one: styled-components, glamorous, etc.

Anyway, this is another issue. Don't have any relation with typescript or this pull request.

File: packages/react-emotion/typings/react-emotion.d.ts (hunk 2 above)
Comment (renatorib) [reply packages/react-emotion/typings/react-emotion.d.ts]: I don't know exactly all edge cases of emotion API. I'll try to study it to improve the coverage support.

File: packages/react-emotion/typescript_tests/tsconfig.json (hunk 1 above)
Comment (renatorib) [reply packages/react-emotion/typescript_tests/tsconfig.json]: Done! Added.
No warnings
:smile:

File: packages/react-emotion/typings/react-emotion.d.ts (hunk 2 above)
Comment (renatorib) [reply packages/react-emotion/typings/react-emotion.d.ts]: @mitchellhamilton I've added support to these two cases you mentioned. 

```tsx
export type InterpolationFn<Props = {}> =
  (props: Props) =>
    | EmotionInterpolation
    | InterpolationFn<Props>

export type InterpolationTypes<Props = {}> =
  | InterpolationFn<Props>
  | EmotionInterpolation

export type Interpolation<Props = {}> =
  | InterpolationTypes<Props>
  | InterpolationTypes<Props>[]
```

If you know more cases, just tell me, please.

**Comment by renatorib** [comment]
Someone can help me with documentation? I think it's the only thing missing to this pr.

My suggestion is to write a section "TypeScript" in emotion website, but you are creating a new website, right? @tkh44

**Comment by tkh44** [comment]
@renatorib any new markdown file in the `/docs` folder is added to the website. Just add a new file there called `typescript.md` and do your thing 👍

**Comment by renatorib** [comment]
That's great! I'll try to write something. :)

**Comment by hemlok** [comment]
This is awesome, I just started using typescript as a replacement for PropTypes, and this will go nicely with that 👌 Thanks @renatorib

**Comment by screendriver** [comment]
> If you don't want to use babel (or can't) just don't use emotion. Pick another one: styled-components, glamor/ous, fela, etc.

I am not complaining about that. It is much more a question how to use emotion with TypeScript 😉 Because you found obviously a way that I am not aware of at the moment and I really want to use emotion as well.

**Comment by renatorib** [comment]
Today I'm 'blocked' from using emotion in my company codebase because we use typescript.
What I did:
- [ ] Ask for a improve on every issue even non-related about ts.
- [x] Sent a PR so I can use it.

If you really needs use emotion and can't use Babel, you can:
1. Use without babel, abdicating the shorthands (using `styled('div')`).
2. Sent a PR that makes emotion shorthands work without babel, so we can work together in that.

I still do not think we should be talking about it here. This has no connection to the PR thread.

**Comment by tkh44** [comment]
@renatorib Is this ready except for documentation?

**Comment by renatorib** [comment]
@tkh44 yep! 

I did not have time to write it yet, but it's the only thing missing.

**Comment by renatorib** [comment]
If you want, you can merge it and I send another PR with documentation.

**Comment by renatorib** [comment]
I've added a typescript documentation

Unfortunately my englis  is not as good as I wish it were.
Someone please give a review so nothing's going wrong :smile: 
Thanks!

**Comment by tkh44** [comment]
Great work @renatorib! I did a quick edit but I'd like someone who knows more about typescript do a review.

**Comment by renatorib** [comment]
Looks great @tkh44, thanks!
